internal/services/elasticsan/elastic_san_volume_resource.go

```diff
@@ -0,0 +1,319 @@
+	VolumeGroupId        string                         `tfschema:"volume_group_id"`
+	VolumeId             string                         `tfschema:"volume_id"`
+}
+
+type ElasticSANVolumeCreateSource struct {
+	SourceType string `tfschema:"source_type"`
+	SourceId   string `tfschema:"source_id"`
+}
+
+func (r ElasticSANVolumeResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return volumes.ValidateVolumeID
+}
+
+func (r ElasticSANVolumeResource) ResourceType() string {
+	return "azurerm_elastic_san_volume"
+}
+
+func (r ElasticSANVolumeResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validate.ElasticSanVolumeName,
+		},
+
+		"volume_group_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: volumes.ValidateVolumeGroupID,
+		},
+
+		"size_in_gib": {
+			Type:         pluginsdk.TypeInt,
+			Required:     true,
+			ValidateFunc: validation.IntBetween(1, 65536),
+		},
+
+		"create_source": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			ForceNew: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"source_id": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						ValidateFunc: validation.Any(
+							commonids.ValidateManagedDiskID,
+							restorepoints.ValidateRestorePointID,
+							diskSnapshots.ValidateSnapshotID,
+							snapshots.ValidateSnapshotID,
+						),
+					},
+					"source_type": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						ValidateFunc: validation.StringInSlice([]string{
+							// None is not exposed
+							string(volumes.VolumeCreateOptionDisk),
+							string(volumes.VolumeCreateOptionDiskRestorePoint),
+							string(volumes.VolumeCreateOptionDiskSnapshot),
+							string(volumes.VolumeCreateOptionVolumeSnapshot),
+						},
+							false),
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r ElasticSANVolumeResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"target_iqn": {
+			Computed: true,
+			Type:     pluginsdk.TypeString,
+		},
+
+		"target_portal_hostname": {
+			Computed: true,
+			Type:     pluginsdk.TypeString,
+		},
+
+		"target_portal_port": {
+			Computed: true,
+			Type:     pluginsdk.TypeInt,
+		},
+
+		"volume_id": {
+			Computed: true,
+			Type:     pluginsdk.TypeString,
+		},
+	}
+}
+
+func (r ElasticSANVolumeResource) CustomizeDiff() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			if oldVal, newVal := metadata.ResourceDiff.GetChange("size_in_gib"); newVal.(int) < oldVal.(int) {
+				return fmt.Errorf("new size_in_gib should be greater than the existing one")
```

Comment:
shall we add a testcase for this validation?

Reply:
This is included in update test